File: leetcode/1001-2000/1201-1300/1268-search-suggestions-system/SearchSuggestionsSystem.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SearchSuggestionsSystem {
  public static void main(String[] args) {
    String[] products = new String[] {
        "mobile", "mouse", "moneypot", "monitor", "mousepad"
    };

    String searchWord = "mouses";

    System.out.println(new Solution().suggestedProducts(products, searchWord));
  }
}

class Trie {
  Trie[] children;
  List<Integer> indices;

  public Trie() {
    children = new Trie[26];
    indices = new ArrayList<>();
  }

  public void populate(String word, int wordIndex) {
    Trie trie = this;

    for (char c : word.toCharArray()) {
      int i = c - 'a';

      if (trie.children[i] == null) {
        trie.children[i] = new Trie();
      }

      if (trie.children[i].indices.size() < 3) {
        trie.children[i].indices.add(wordIndex);
      }

      trie = trie.children[i];
    }
  }

  public List<List<Integer>> search(String word) {
    List<List<Integer>> answers = new ArrayList<>();

    Trie trie = this;

    for (char c : word.toCharArray()) {
      int index = c - 'a';

      if (trie == null || trie.children[index] == null) {
        return answers;
      }

      answers.add(trie.children[index].indices);

      trie = trie.children[index];
    }

    return answers;
  }
};

class Solution {
  public List<List<String>> suggestedProducts(String[] products, String searchWord) {
    Trie trie = new Trie();

    Arrays.sort(products);

    for (int i = 0; i < products.length; i++) {
      trie.populate(products[i], i);
    }

    List<List<Integer>> indices = trie.search(searchWord);
    List<List<String>> answers = new ArrayList<>();

    for (List<Integer> integers : indices) {
      List<String> l = new ArrayList<>();
      for (Integer i : integers) {
        l.add(products[i]);
      }

      answers.add(l);
    }

    while (answers.size() < searchWord.length()) {
      answers.add(new ArrayList<>());
    }

    return answers;
  }
}

File: leetcode/1001-2000/1201-1300/1261-1270/1268-search-suggestions-system/SearchSuggestionsSystem.ts
class Trie {
  children: (Trie | null)[];
  indices: number[];

  constructor() {
    this.children = new Array(26).fill(null);
    this.indices = [];
  }

  populate(s: string, index: number) {
    let trie: Trie = this;

    for (let i = 0; i < s.length; i++) {
      const c = s.charCodeAt(i) - 'a'.charCodeAt(0);

      const child = trie.children[c] ?? new Trie();

      if (child.indices.length < 3) {
        child.indices.push(index);
      }

      trie.children[c] = child;

      trie = child;
    }
  }

  search(s: string): number[][] {
    let trie: Trie = this;
    const answers: number[][] = [];

    for (let i = 0; i < s.length; i++) {
      // const c = s[i];
      const c = s.charCodeAt(i) - 'a'.charCodeAt(0);

      // console.log(c, JSON.stringify(trie));

      if (trie == null || trie.children[c] == null) {
        return answers;
      }

      answers.push(trie.children[c].indices);

      trie = trie.children[c];
    }

    return answers;
  }
};

function suggestedProducts(products: string[], searchWord: string): string[][] {
  let trie: Trie | undefined | null = new Trie();
  products.sort((a, b) => a > b ? 1 : -1);
  products.forEach((p, i) => trie?.populate(p, i));

  // console.log(JSON.stringify(trie));

  const words = trie.search(searchWord).map(
    (indies) => indies.map(i => products[i])
  );

  while (words.length < searchWord.length) {
    words.push([]);
  }

  return words;
};

// console.log(suggestedProducts(["mo", "ma"], "mas"));
console.log(suggestedProducts(["mobile", "mouse", "moneypot", "monitor", "mousepad"], "mouse"));
